server/src/main/java/io/druid/segment/indexing/DataSchema.java

```diff
@@ -94,53 +138,75 @@ public DataSchema(
                   )
         );
       } else {
-        this.parser = parser;
+        return inputRowParser;
       }
     } else {
-      log.warn("No parser or parseSpec has been specified");
-
-      this.parser = parser;
-    }
-
-    if (aggregators.length == 0) {
-      log.warn("No metricsSpec has been specified. Are you sure this is what you want?");
+      log.warn("No parseSpec in parser has been specified.");
+      return inputRowParser;
     }
-    this.aggregators = aggregators;
+  }
 
-    if (granularitySpec == null) {
-      log.warn("No granularitySpec has been specified. Using UniformGranularitySpec as default.");
-      this.granularitySpec = new UniformGranularitySpec(null, null, null);
-    } else {
-      this.granularitySpec = granularitySpec;
-    }
+  @JsonProperty("metricsSpec")
+  public AggregatorFactory[] getAggregators()
+  {
+    return aggregators;
   }
 
   @JsonProperty
-  public String getDataSource()
+  public GranularitySpec getGranularitySpec()
   {
-    return dataSource;
+    return granularitySpec;
   }
 
-  @JsonProperty
-  public InputRowParser getParser()
+  public DataSchema withGranularitySpec(GranularitySpec granularitySpec)
   {
-    return parser;
+    return new DataSchema(dataSource, parser, aggregators, granularitySpec, jsonMapper);
   }
 
-  @JsonProperty("metricsSpec")
-  public AggregatorFactory[] getAggregators()
+  @Override
+  public boolean equals(Object o)
   {
-    return aggregators;
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    DataSchema that = (DataSchema) o;
+
+    if (!dataSource.equals(that.dataSource)) {
+      return false;
+    }
+    if (parser != null ? !parser.equals(that.parser) : that.parser != null) {
+      return false;
+    }
+    // Probably incorrect - comparing Object[] arrays with Arrays.equals
```

Comment:
equals/hashCode are generated automatically by intellij and we generally dont touch what is produced. that comment is weird but I left it to follow "not editting generated code" tradition.